File: Palindrome.java
// Overall Time Complexity : O(n) . The n is number of characters in  the string.
// Overall Space Complexity : O(1) . The space complexity of storing alphabets is 0(1).
// Any problem you faced while coding this : No issues faced during implementation


// The solution to the problem involves creating a HashSet and checking if the occurence of a character is more than once.

import java.util.*;
public class Palindrome {
    public int longestPalindrome(String s) {
        HashSet<Character> hset = new HashSet<>();
        int res = 0;

        for(int i =0; i<s.length(); i++){
            char ch = s.charAt(i);
            if(!hset.contains(ch)){
                hset.add(ch);
            }
            else{
                res +=2;
                hset.remove(ch);
            }

        }
        if(!hset.isEmpty()){
            res+=1;
        }
        return res;
    }
}

File: palindrome.py
'''
Time Complexity :
O(n)
Space Complexity : O(n))
Did this code successfully run on Leetcode : yes
Any problem you faced while coding this : No
'''
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        hash1 = {}
        for c in s:
            if c not in hash1:
                hash1[c]=0
            hash1[c]+=1

        count=0
        odds=0
        odd_found=False
        for k, v in hash1.items():
            if v%2!=0:
                count+=v-1
                odd_found=True
            elif v%2==0:
                count+=v
        
        if odd_found:
            count=count+1
        
        return count